dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -170,16 +177,32 @@ StableValueSpacePtr createNewStable( //
     PageIdU64 stable_id,
     WriteBatches & wbs)
 {
-    auto delegator = context.path_pool.getStableDiskDelegator();
+    auto delegator = context.path_pool->getStableDiskDelegator();
     auto store_path = delegator.choosePath();
 
-    PageIdU64 dtfile_id = context.storage_pool.newDataPageIdForDTFile(delegator, __PRETTY_FUNCTION__);
+    PageIdU64 dtfile_id = context.storage_pool->newDataPageIdForDTFile(delegator, __PRETTY_FUNCTION__);
     auto dtfile = writeIntoNewDMFile(context, schema_snap, input_stream, dtfile_id, store_path);
 
     auto stable = std::make_shared<StableValueSpace>(stable_id);
     stable->setFiles({dtfile}, RowKeyRange::newAll(context.is_common_handle, context.rowkey_column_size));
     stable->saveMeta(wbs.meta);
-    wbs.data.putExternal(dtfile_id, 0);
+    if (auto data_store = context.db_context.getRemoteDataStore(); !data_store)
+    {
+        wbs.data.putExternal(dtfile_id, 0);
+    }
+    else
+    {
+        auto store_id = context.db_context.getTMTContext().getKVStore()->getStoreID();
+        // FIXME: get table_id from context
+        Remote::DMFileOID oid{.store_id = store_id, .table_id = 0, .file_id = dtfile_id};
+        data_store->putDMFile(dtfile, oid);
+        PS::V3::CheckpointLocation loc{
+            .data_file_id = std::make_shared<String>(S3::S3Filename::fromDMFileOID(oid).toFullKey()),
```

Comment:
Do we need to add a field to distinguish the DTFile data (external data) and the CheckpointFile data?

Reply:
Seems not needed since they are the same "DataFile" at the point of view from S3
https://github.com/pingcap/tiflash/blob/ce007dbe642abb4074eb3c83ce0b28a7078fdc8e/dbms/src/Storages/S3/S3Filename.h#L46

Reply:
What do you think? /cc @lidezhu

Reply:
I think it's ok to reuse this field. Add another field doesn't make the concept of `CheckpointLocation` more clear and add some more complexity.